### PR TITLE
Add type assertion linter

### PR DESCRIPTION
**Pull request by dmytzo**

### CODE REVIEW

File: linters/linters.go [hunk 1]
```diff
@@ -90,3 +91,31 @@ func channelSize(pass *analysis.Pass) (interface{}, error) {
 
 	return nil, nil
 }
+
+func typeAssertion(pass *analysis.Pass) (interface{}, error) {
+	for _, file := range pass.Files {
+		ast.Inspect(file, func(n ast.Node) bool {
+			as, ok := n.(*ast.AssignStmt)
```
Comment (iorvd) [inline linters/linters.go:98]: I think we don't need to limit the possible use to assignment only, let's expand it for all possible type assertions.

So we have the following cases here and we need process them all properly:
1. `x, ok := y.(type)` - VALID
2. `x, _ := y.(type)` - INVALID
3. `f(y.(type))` - INVALID
4. `return y.(type)` - INVALID
5. ...

There might be a lot of different options, so I suggest to simplify: the 1st step is to check whether it's a type assertion node or not, the 2nd is to go back in AST and make sure it's a safe type assertion.

WDYT?

**Review by iorvd — CHANGES_REQUESTED** [review]

File: linters/linters.go (hunk 1 above)
Comment (dmytzo) [reply linters/linters.go:98]: thx! good idea, I will try

File: linters/linters.go [hunk 2]
```diff
@@ -1,121 +1,148 @@
...
-				return false
-			}
+	as, ok := nodes[len(nodes)-2].(*ast.AssignStmt)
+	if !ok {
+		return err
+	}
 
-			return true
-		})
+	if len(as.Lhs) != 2 || as.Lhs[1].(*ast.Ident).String() == _BlankIdentifier {
```
Comment (iorvd) [inline linters/linters.go:143]: This linter wouldn't pass a run against itself :)

File: linters/testdata/src/typeassertion/main.go [hunk 3]
```diff
@@ -1,13 +1,20 @@
...
+	f(i.(string))                         // want "Handle type assertion failures"
+	func() string { return i.(string) }() // want "Handle type assertion failures"
+	_ = ff{i.(string)}                    // want "Handle type assertion failures"
+	_ = ff{s: i.(string)}                 // want "Handle type assertion failures"
+
 	c, ok := i.(string)
 
-	fmt.Println(a, b, c, ok)
+	_, _, _, _ = a, b, c, ok
 }
```
Comment (iorvd) [inline linters/testdata/src/typeassertion/main.go:20]: awesome!🕺

File: linters/linters.go (hunk 2 above)
Comment (dmytzo) [reply linters/linters.go]: I know :)

Updated

File: linters/linters.go [hunk 4]
```diff
@@ -1,121 +1,148 @@
...
+				pass.Reportf(lErr.pos, lErr.Error())
+				return false
+			}
 		}
+
+		return true
 	}
+}
+
+func noinit(_ *analysis.Pass, file *ast.File, _ []ast.Node, n ast.Node) error {
```
[the remark this answers is not in the document]
Comment (dmytzo) [reply linters/linters.go:75]: I do not think we need to do it already. We can try, when we have some big linters with additional funcs

**Review by iorvd — APPROVED** [review]

**Pull request merged**